### PR TITLE
Add PF2e ranged Strike support

### LINKED ISSUE #76: [Combat] Add full PF2e ranged Strike support
## Context
A targeting/grid audit found that Strike support is currently melee-focused and cannot execute official Pathfinder 2e ranged Strikes. This blocks future player and enemy use of bows, slings, thrown weapons, and other ranged attacks.

Rules references:
- Strike: https://2e.aonprd.com/Actions.aspx?ID=2306
- Attack rolls, MAP, and range penalties: https://2e.aonprd.com/Rules.aspx?ID=2288
- Range and reach: https://2e.aonprd.com/Rules.aspx?ID=2379
- Grid movement and cover: https://2e.aonprd.com/Rules.aspx?ID=2345
- Line of effect: https://2e.aonprd.com/Rules.aspx?ID=2382

## Current Behavior
- `WeaponStrikeAdderAutomatic` only adds the first listed melee weapon where `weapon.range == 0`, so weapons like Shortbow are skipped.
- `GridAPI.GetStrikeTarget` and `StateStrike` only select one occupant in a highlighted radius.
- `Dijkstra.CalculateEmination` is used as a generic range helper, but it is misnamed and uses Euclidean distance rather than PF2e grid measurement.
- There is no active line-of-effect/line-of-sight enforcement; `LineOfSight.cs` is commented out.
- Range increments, cover, ammunition, reload, ranged damage rules, and action-specific attack bonuses are not handled.
- `D20.Roll` treats natural 20 as automatic critical success and natural 1 as automatic critical failure instead of PF2e's one-degree adjustment.

## Expected Behavior
Both player characters and enemies can use official PF2e ranged Strikes through the same action pipeline as melee Strikes.

## Scope
- Add a reusable creature-targeting request/result model for Strike-style targeting, including selected target, distance, line-of-effect result, cover result, and range penalty.
- Add PF2e grid range measurement using 5-foot squares and alternating 5/10-foot diagonals for targeting. Do not refactor existing Stride pathfinding unless tests prove it is required for this issue.
- Enforce ranged Strike range increments: first increment no penalty, then `-2` per additional increment, max `-10`, impossible beyond six increments.
- Apply multiple attack penalty and range penalties together in Strike resolution.
- Add active line-of-effect and solid-wall line-of-sight checks using current grid tile data.
- Apply cover modifiers to ranged attack AC where current terrain supports standard or greater cover.
- Add all valid melee and ranged weapon Strike actions without duplicates.
- Add minimal ammo/reload state: projectile weapons require ammo, reload `0` weapons can fire if ammo exists, reload `>0` weapons become unloaded after firing and require a reload action with the weapon's reload cost.
- Correct ranged damage profiles: bows do not add melee flat damage; implement required local ranged traits such as `deadly-d10`.
- Load or derive action-specific attack modifiers where available, especially imported NPC weapon action bonuses.
- Update AI selection so enemies can choose legal ranged Strikes instead of always moving adjacent.

## Acceptance Criteria
- [ ] PCs and enemies receive ranged weapon Strike actions when equipped or listed with ranged weapons.
- [ ] A Shortbow Strike can target creatures within six 60-foot increments and applies the correct range penalty.
- [ ] A target behind a blocking wall cannot be selected because there is no line of effect.
- [ ] Cover modifies ranged attack AC according to PF2e rules where terrain partially blocks the line.
- [ ] MAP still increments after Strike and combines correctly with range penalty.
- [ ] Bow damage does not add melee flat damage, and `deadly-d10` applies only on critical success.
- [ ] Ammo is consumed and reload state is enforced for ranged projectile weapons.
- [ ] AI can choose a legal ranged Strike when it has a ranged weapon and target in range.
- [ ] EditMode tests cover d20 degree adjustment, range increments, penalties, ranged damage, ammo/reload, and line-of-effect legality.
- [ ] PlayMode tests cover player targeting/cancel behavior and enemy ranged Strike use.
- [ ] Unity EditMode and PlayMode test commands from `AGENTS.md` pass without `-quit`.

## Notes
Likely files/subsystems:
- `Assets/Scripts/Interfaces/GridAPI.cs`
- `Assets/Scripts/Grid/FSM/StateStrike.cs`
- `Assets/Scripts/Grid/Dijkstra.cs`
- `Assets/Scripts/Combat/Actions/Attacks/Strike.cs`
- `Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs`
- `Assets/Scripts/Creature/Dice.cs`
- `Assets/Scripts/Creature/CreatureComponent.cs`
- `Assets/Scripts/Combat/MindlessController.cs`

Create implementation work in a task worktree under `../sdmay26-02-worktrees/`.

### PR DESCRIPTION
**Pull request by clausman**
Closes #76.

## Summary
- Add Strike targeting requests/results with PF2e grid distance, range increments, line of effect, and cover.
- Add ranged weapon Strike actions, action-specific attack bonuses, ammo/reload state, ranged damage handling, and deadly-d10 critical damage.
- Update AI to choose legal ranged Strikes before moving adjacent.

## Tests
- Unity EditMode: 8/8 passed.
- Unity PlayMode: 25/25 passed.

### CODE REVIEW

File: Assets/Scripts/Combat/Actions/Attacks/Strike.cs [hunk 1]
```diff
@@ -1,102 +1,139 @@
...
-        }
+        uint strikePenaltyCount = From.GetComponent<ActionController>()?.StrikePenalty ?? 0;
+        int mapPenalty = CalculateMultipleAttackPenalty(strikePenaltyCount);
+        int rangePenalty = TargetingResult?.RangePenalty ?? 0;
+        int coverBonus = TargetingResult?.CoverAcBonus ?? 0;
 
-        // TODO calculate actual bonuses
-        int attackBonus = from.attackBonus;
-        int damageBonus = from.damageBonus;
+        int attackBonus = AttackBonusOverride ?? from.attackBonus;
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/Strike.cs]: What is the purpose of AttackBonusOverride here? In general, Pathfinder has four types of bonuses to values: Flat, Item, Status, and Circumstance. Item, Status, and Circumstance always take the highest bonus/penalty type and don't stack. Flat bonuses do stack and almost always come from feats (e.g. the Fleet feat's boost to speed).
This is important to treat correctly, because Flat-Footed is a circumstance penalty to you AC. So some other effect that gives a status penalty to your AC would NOT stack with Flat-Footed, the higher would apply. Same applies to Cover, it gives a circumstance bonus to AC. So both Cover would not stack with other circumstance bonuses to AC, but both it and a Status bonus to AC could apply.
I think you want to model this correctly when calculating bonuses here to future proof the calculation.

File: Assets/Scripts/Combat/Actions/Attacks/Strike.cs [hunk 2]
```diff
@@ -1,102 +1,139 @@
...
 
         if (attackRoll.degree == DegreeOfSuccess.Success || attackRoll.degree == DegreeOfSuccess.CriticalSuccess)
         {
             CombatLog.GetInstance().Log(log);
             OnDamageDealt.Invoke(Damages[0].damageType);
-            // Adds a new flat damage for the damage bonus, type matching the first damage type
-            // FlatDamages.Add(new DamageValue(Damages[0].damageType, damageBonus));
             List<DamageValue> damageValues = DamageRoller.RollDamage(Damages, FlatDamages);
             DamageRoller.EvaluateCriticalDamage(attackRoll.degree, damageValues);
+            ApplyDeadlyDamage(attackRoll.degree, damageValues);
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/Strike.cs:79]: There are other crit effects that happen in pf2e we will need to support in the future, like the fatal trait and even Weapon Specialization Effects. Plus certain monsters like Constructs have additional effects when they are Crit. Does this design lend itself well to those future extensions or can it be designed better to make it more flexible in the future?

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs [hunk 3]
```diff
@@ -3,66 +3,83 @@
...
         CreatureComponent cc = creature.GetComponent<CreatureComponent>();
-        // Check each hand for a weapon, and add corresponding strike action
-        // Assumes that CreatureComponent is properly enforcing rules for what can be equipped
-        // TODO prevent duplicate actions if called multiple times
+        if (cc == null)
+            return;
+
         if (cc.HasEquippedRightWeapon())
+            AddWeaponStrike(creature, cc.equippedRightHand);
+        if (cc.HasEquippedLeftWeapon())
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs:30]: How does this work with creatures who don't have two hands? e.g. a bite, claws, and tail attack?

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs [hunk 4]
```diff
@@ -3,66 +3,83 @@
...
 
 [System.Serializable]
 public class StrikeWeapon : MultiFrameEntityAction
 {
-    // Done by Ryan Meyer 04/07/2026
     public override string ActionName => GetWeaponName();
     private Strike Strike;
     private EquipmentWeapon Weapon;
-    private int range =1; // default range of 1 tile
+    private int range = 1;
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Since the Weapon is already referenced in this class, can its range be determined from the weapon itself by checking for the Reach trait and its associated value?

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs [hunk 5]
```diff
@@ -3,66 +3,83 @@
...
+            return;
+
+        ActionController controller = creature.GetComponent<ActionController>();
+        CreatureComponent cc = creature.GetComponent<CreatureComponent>();
+        if (controller == null || cc == null || HasAction(controller, weapon.name))
+            return;
+
+        StrikeWeapon strikeWeapon = new StrikeWeapon(1, weapon, creature);
+        controller.AddAction(strikeWeapon);
+        if (!cc.actions.Contains("StrikeWeapon " + strikeWeapon.weaponName))
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Are these hard-coded string constants being used to determine business logic? If so, please clean them up. As I'm sure you know checking existence or identify with strings isn't great for maintainablity

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs [hunk 6]
```diff
@@ -83,57 +100,121 @@ public Strike GetStrike()
...
+        ActionController ac = target.GetComponent<ActionController>();
+        CreatureComponent cc = target.GetComponent<CreatureComponent>();
+        if (cc != null && cc.ReloadWeapon(Weapon))
+        {
+            CombatLog.GetInstance().Log("- " + target.name + " reloads " + Weapon.name + ".");
+            PayCost(ac);
+        }
+        if (ac)
+            ac.IsTakingAction = false;
+        CombatManager.GetInstance().CheckForEndOfGame();
```
Comment (clausman) [inline Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Why is this `CheckForEndOfGame();` needed here? It seems odd that the Reload action would check for the end of the game. Is this an existing code smell in the repo?

File: Assets/Scripts/Grid/States/StateStrike.cs [hunk 7]
```diff
@@ -1,16 +1,17 @@
...
     public class StateStrike : GridFSMState
     {
         protected GridFSM Fsm;
-        // target character
-        GameObject Character;
-        CoroutineResult<GameObject> Selection;
+        private readonly GameObject Character;
+        private readonly StrikeTargetRequest Request;
+        private readonly CoroutineResult<StrikeTargetResult> Selection;
+        private readonly CoroutineResult<GameObject> LegacySelection;
```
Comment (clausman) [inline Assets/Scripts/Grid/States/StateStrike.cs]: What is the rationale for both Selection and LegacySelection here? Keep in mind that backwards compatability does NOT matter right now since this game is still in development.

File: Assets/Scripts/Grid/GridBase.cs [hunk 8]
```diff
@@ -84,7 +84,22 @@ public override IEnumerator Stride(GameObject character)
 
         public override IEnumerator GetStrikeTarget(GameObject attacker, float range, CoroutineResult<GameObject> target)
         {
-            if (Fsm.ChangeState(new StateStrike(attacker, range, target, Fsm)))
+            StrikeTargetRequest request = new StrikeTargetRequest
+            {
+                ReachFeet = Mathf.RoundToInt(range),
+                IsRanged = false,
+                RequiresLineOfEffect = true
```
Comment (clausman) [inline Assets/Scripts/Grid/GridBase.cs]: Make sure there are unit tests which verify this works correctly, especially edge cases like how striking "around a corner" works. PF2e should have precise rules for this, and I believe they are something like
1) Can any corner reach any other corner? If yes, valid target, if no, invalid target
2) Can any corner reach all other corners? If yes, no cover for target, if no, partial cover for target
Although if you don't have cover implemented yet, that's okay. Just make sure what is implemented is correct.

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs (hunk 6 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Yes, this is an existing action-contract smell. `EntityAction.Invoke()` performs the end-of-game check, and `MultiFrameEntityAction` does the same after its coroutine. Since `ReloadWeaponAction` overrides `Invoke()` directly, I originally mirrored that check manually.

I pushed `5c8a0c4` to delegate to `base.Invoke(target)` instead. That keeps reload aligned with the existing action contract and removes the direct `CombatManager.GetInstance().CheckForEndOfGame()` call from reload.

File: Assets/Scripts/Combat/Actions/Attacks/Strike.cs (hunk 1 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/Strike.cs]: Good call. `AttackBonusOverride` was poorly named because it sounded like a PF2e bonus type when it is really the total imported NPC weapon action modifier. I pushed `5c8a0c4` to rename it to `AttackModifierOverride` so it is not confused with item/status/circumstance/untyped bonus stacking.

I agree the real future-proof model should be typed modifiers with PF2e stacking rules. I did not fold that larger system into this PR, but captured it as #81 so cover, off-guard/flat-footed, status bonuses, item bonuses, etc. can all go through one resolver.

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs (hunk 4 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Yes. I pushed `5c8a0c4` to remove the cached `range` field from `StrikeWeapon`. `GetRange()` now derives from `GetTargetRequest()`, and the request derives melee reach/ranged range from the weapon each time.

One caveat: current weapon data only gives us a `reach` trait string, so this still treats reach as 10 ft. If imported data later distinguishes larger reach values, the next step is parsing that into structured weapon/reach data instead of treating `reach` as a boolean trait.

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs (hunk 5 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs]: Agreed. I pushed `5c8a0c4` so duplicate/action identity checks no longer compare action-name strings. They now use typed checks: `StrikeWeapon` matches by represented weapon and `ReloadWeaponAction` does the same.

The strings that remain here are UI/display labels or raw imported data identifiers (`reach`, `reload-*`, ammo names, etc.). Those are still string-backed because the current JSON/import model is string-backed, but they are not used for duplicate action identity anymore.

File: Assets/Scripts/Grid/States/StateStrike.cs (hunk 7 above)
Comment (clausman) [reply Assets/Scripts/Grid/States/StateStrike.cs]: Agreed. I added that bridge only to keep the old target API alive while wiring ranged targeting, but we do not need that compatibility during development.

I pushed `5c8a0c4` to remove `LegacySelection`, remove the old `GridAPI.GetStrikeTarget(... CoroutineResult<GameObject>)` overload, and update unarmed strikes to use the same `StrikeTargetRequest` / `StrikeTargetResult` path as weapon strikes.

File: Assets/Scripts/Grid/GridBase.cs (hunk 8 above)
Comment (clausman) [reply Assets/Scripts/Grid/GridBase.cs]: The implementation follows that corner-ray shape: it samples 16 corner-to-corner rays. Zero clear rays means no line of effect / illegal target; 16 clear rays means no cover; any value from 1 to 15 currently maps to standard cover.

There is EditMode coverage for both sides of that behavior in `Pf2eRangedStrikeTests`: `SolidWallBlocksLineOfEffect` verifies no clear corner path blocks targeting, and `PartialWallGivesStandardCover` verifies an around-corner/partial line gives cover. The final test runs after the cleanup are still green: EditMode 8/8 and PlayMode 25/25.

File: Assets/Scripts/Combat/Actions/Attacks/Strike.cs (hunk 2 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/Strike.cs:79]: This PR's `deadly-d10` support is intentionally local and not the final shape for every critical effect. It is enough to make shortbow damage correct for this issue, but it is not the right long-term extension point for fatal, critical specialization, or defender-side critical effects.

I captured that as #80. The design I would move toward there is a post-roll attack result context containing attacker, target, strike, traits/weapon, degree of success, damage values, and targeting result, then let critical-only traits and defender effects subscribe/evaluate from that context rather than adding each rule directly inside `Strike.DamageEvaluate`.

File: Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs (hunk 3 above)
Comment (clausman) [reply Assets/Scripts/Combat/Actions/Attacks/StrikeWeapon.cs:30]: The equipped-hand path (`WeaponStrikeAdder`) is still hand-slot based, so it is mostly for PC/equipment-style weapons. For imported/listed creature weapons, `WeaponStrikeAdderAutomatic` adds every weapon in `CreatureComponent.weapons` / `weaponsList` regardless of hand slots, so it does not require a creature to have two hands.

That said, true natural attacks like bite/claw/tail are not represented cleanly yet; they are closer to creature action entries than equipment. This PR does not solve the natural-attack data model, it just avoids blocking ranged weapon actions. We should model natural attacks separately from hand equipment when we flesh out NPC action import further.
